File: Java-Core/src/main/java/presto/statement/DeclarationInstruction.java
package presto.statement;

import presto.declaration.ConcreteMethodDeclaration;
import presto.declaration.IDeclaration;
import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.runtime.Context;
import presto.runtime.Variable;
import presto.type.IType;
import presto.type.MethodType;
import presto.type.VoidType;
import presto.utils.CodeWriter;
import presto.value.IValue;
import presto.value.ClosureValue;

public class DeclarationInstruction<T extends IDeclaration> extends BaseStatement {

	T declaration;
	
	public DeclarationInstruction(T declaration) {
		this.declaration = declaration;
	}
	
	public T getDeclaration() {
		return declaration;
	}
	
	@Override
	public void toDialect(CodeWriter writer) {
		if(declaration instanceof ConcreteMethodDeclaration) try {
			ConcreteMethodDeclaration method = (ConcreteMethodDeclaration)declaration;
			writer.getContext().registerDeclaration(method);
		} catch(SyntaxError e) {
			// ok
		}
		declaration.toDialect(writer);
	}
	
	@Override
	public IType check(Context context) throws SyntaxError {
		if(declaration instanceof ConcreteMethodDeclaration) {
			ConcreteMethodDeclaration method = (ConcreteMethodDeclaration)declaration;
			method.checkChild(context);
			context.registerDeclaration(method);
		} else
			throw new SyntaxError("Unsupported:" + declaration.getClass().getSimpleName());
		return VoidType.instance();
	}
	
	@Override
	public IValue interpret(Context context) throws PrestoError {
		if(declaration instanceof ConcreteMethodDeclaration) {
			ConcreteMethodDeclaration method = (ConcreteMethodDeclaration)declaration;
			context.registerDeclaration(method);
			IType type = new MethodType(context,method.getName());
			context.registerValue(new Variable(method.getName(), type)); 
			context.setValue(method.getName(), new ClosureValue(context, method));
			return null;
		} else
			throw new SyntaxError("Unsupported:" + declaration.getClass().getSimpleName());
	}
	
}

File: Java-CodeServer/src/main/resources/js/prompto/statement/DeclarationInstruction.js
var ConcreteMethodDeclaration = require("../declaration/ConcreteMethodDeclaration").ConcreteMethodDeclaration;
var ExpressionValue = require("../value/ExpressionValue").ExpressionValue;
var ClosureValue = require("../value/ClosureValue").ClosureValue;
var BaseStatement = require("./BaseStatement").BaseStatement;
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var MethodType = require("../type/MethodType").MethodType;
var VoidType = require("../type/VoidType").VoidType;
var Variable = require("../runtime/Variable").Variable;

function DeclarationInstruction(declaration) {
	BaseStatement.call(this);
	this.declaration = declaration;
	return this;
}

DeclarationInstruction.prototype = Object.create(BaseStatement.prototype);
DeclarationInstruction.prototype.constructor = DeclarationInstruction;

DeclarationInstruction.prototype.toDialect = function(writer) {
    if(this.declaration instanceof ConcreteMethodDeclaration) try {
        writer.context.registerMethodDeclaration(this.declaration);
     } catch(e) {
        // ok
     }
    this.declaration.toDialect(writer);
};


DeclarationInstruction.prototype.check = function(context) {
	if(this.declaration instanceof ConcreteMethodDeclaration) {
		this.declaration.checkChild(context);
		context.registerMethodDeclaration(this.declaration);
	} else {
		throw new SyntaxError("Unsupported:" + typeof(declaration));
	}
	return VoidType.instance;
};

DeclarationInstruction.prototype.interpret = function(context) {
	if(this.declaration instanceof ConcreteMethodDeclaration) {
		var method = this.declaration;
		context.registerMethodDeclaration(method);
		var type = new MethodType(context,method);
		context.registerValue(new Variable(method.name, type));
		context.setValue(method.name, new ClosureValue(context, method));
		return null;
	} else {
		throw new SyntaxError("Unsupported:" + typeof(this.declaration));
	}
};


exports.DeclarationInstruction = DeclarationInstruction;
